### PR TITLE
Encrypt local API key

### LINKED ISSUE #219: Consider encypting the locally stored API key with user password

### PR DESCRIPTION
**Pull request by ericmj**
I would like feedback on the general idea of encrypting the API key. This makes it harder to steal your key and stops the NPM worm vulnerability [1] reported a while back (we still need 2-factor though). The downside is that you always have to enter your password when publishing packages, adding/removing package owners and so on.

I would also like feedback on the encryption scheme. AES-CBC is used to encrypt the key. PBKDF2 is used for key stretching with 32768 iterations to make bruteforcing slower (on my late 2013 macbook pro stretching takes ~100ms). A 16 byte salt is generated and stored with the encrypted key. The string "HEXAPIKEY" is prepended to the plaintext which along with the padding is used to detect wrong passwords.

If an unencrypted key is detected in the config we will force the user to encrypt it. When you fetch a new key from the API server it will be encrypted with your account password.

Closes #219.

[1] https://www.kb.cert.org/vuls/id/319816

### CODE REVIEW

File: lib/hex/state.ex [hunk 1]
```diff
@@ -46,7 +46,9 @@ defmodule Hex.State do
       hexpm_pk:         @hexpm_pk,
       registry_updated: false,
       httpc_profile:    :hex,
-      ssl_version:      ssl_version()}
+      ssl_version:      ssl_version(),
+      pbkdf2_iters:     32768,
```
Comment (jcspencer) [inline lib/hex/state.ex]: This is also defined in `Hex.Crypto`

File: lib/hex/crypto.ex [hunk 2]
```diff
@@ -0,0 +1,59 @@
+defmodule Hex.Crypto do
+  import Hex.Crypto.Pbkdf2
+  @pbkdf2_iterations 32768
+
+  def encrypt(password, salt, plain, tag \\ "") do
+    <<key::binary-32, iv::binary-16>> = pbkdf2(password, salt, iterations(), 48, :sha256)
+    plain = pad(tag <> plain, 16)
+    :crypto.block_encrypt(:aes_cbc, key, iv, plain)
```
Comment (jcspencer) [inline lib/hex/crypto.ex]: This needs to be changed to `:aes_cbc` for everything to work correctly

Comment (ericmj) [inline lib/hex/crypto.ex]: Do you mean `:aes_cbc128`? Is it needed for < OTP 19 compatability?

**Comment by jcspencer** [comment]
LGTM 👍

**Pull request merged**